### DISCUSSION TITLE
Syscalls for light client verification & cross-chain communication

### DISCUSSION THREAD

**Discussion by PhilippeMts** [filecoin-project/FIPs#517] [Filecoin Virtual Machine]
## Context

Cross-network interoperability is one of the applications opened by the new capabilities of the FVM.

Three types of strategies are often discussed to achieve such interoperability[^1][^2]: (1) notary schemes, (2) hash-locking and (3) relays.

**Relays can be very versatile and promising**[^3], especially when combined with multi-layer architectures based on a *relay chain*, like it could be with Cosmos and Polkadot.

Technically, this strategy basically asks for one chain to verify itself information it has been provided about another chain. It is based on *light client verification* which involves two kinds of verifications: (1) consensus verification and (2) transaction inclusion verification.
With regard to the FVM, this means that **possibly only a few cryptographic primitives** (hash functions and signature verification) **could be missing** in order to create relays for other large networks.

## Subjects of discussion

Enticed by the promise of such relays, we seek here to **open the important questions related to the FVM set of syscalls** so that we do not miss these opportunities.

- What set of primitives might be missing?

  > Other hash functions (*eg* in the SHA family) could open the door to communication with interesting networks. Some signature algorithms (*eg* Schnorr) could also be worth considering.

- What should be the interface for new syscalls?

  > Should we dedicate syscalls to direct access to the new crypto primitives, or should we focus on higher level interfaces for batched calls and higher level verifications (consensus and transaction inclusion).

- Are dedicated syscalls required and justified?

  > In which *logical domain* of the FVM should these primitives be executed? This question is closely related to the above one, to ensure performance and avoid repeated boundary traversals.

> We are grateful to the members of the *FVM // Cross-chain* working group whose work has made this discussion possible. 🙏

[^1]: Robinson, Peter. "Survey of crosschain communications protocols." *Computer Networks* 200 (2021): 108488.
[^2]: Buterin, Vitalik. "Chain interoperability." *R3 Research Paper* 9 (2016).
[^3]: Zamyatin, Alexei, et al. "Xclaim: Trustless, interoperable, cryptocurrency-backed assets." *2019 IEEE Symposium on Security and Privacy (SP)*. IEEE, 2019.

**Comment by anorth** [comment]
Thanks very much for opening this discussion. It might also be (more?) relevant in https://github.com/filecoin-project/FIPs/discussions, especially if the initial milestones are focussed on the transition to WASM rather than adding new functionality.

**Comment by PhilippeMts** [comment]
Thanks @anorth for the advice.

I guess once the discussion here moves from technical notes on the FVM primitives in regards to cross-chain communication, towards something more substantial and beyond the scope of the FVM only, it will indeed make sense to open a thread on the FIPs repository.
